source/snapshot/fetch_worker.go

```diff
@@ -0,0 +1,393 @@
+		if _, err := tx.Exec(ctx, "CLOSE "+f.cursorName); err != nil {
+			sdk.Logger(ctx).Warn().
+				Err(err).
+				Msgf("unexpected error when closing cursor %q", f.cursorName)
+		}
+	}, nil
+}
+
+func (f *FetchWorker) updateFetchLimit(ctx context.Context, tx pgx.Tx) error {
+	if f.snapshotEnd > 0 {
+		return nil
+	}
+
+	if err := tx.QueryRow(
+		ctx,
+		fmt.Sprintf("SELECT max(%s) FROM %s", f.conf.Key, f.conf.Table),
+	).Scan(&f.snapshotEnd); err != nil {
+		return fmt.Errorf("failed to query max on %q.%q: %w", f.conf.Table, f.conf.Key, err)
+	}
+
+	return nil
+}
+
+func (f *FetchWorker) fetch(ctx context.Context, tx pgx.Tx) (int, error) {
+	rows, err := tx.Query(ctx, fmt.Sprintf("FETCH %d FROM %s", f.conf.FetchSize, f.cursorName))
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch rows: %w", err)
+	}
+	defer rows.Close()
+
+	var fields []string
+	for _, f := range rows.FieldDescriptions() {
+		fields = append(fields, f.Name)
+	}
+
+	var nread int
+
+	for rows.Next() {
+		values, err := rows.Values()
+		if err != nil {
+			return 0, fmt.Errorf("failed to get values")
+		}
+
+		if err := f.send(
+			ctx,
+			f.buildRecord(fields, values),
+		); err != nil {
+			return nread, fmt.Errorf("failed to send record")
+		}
+
+		nread++
+	}
+
+	return nread, nil
+}
+
+func (f *FetchWorker) send(ctx context.Context, r sdk.Record) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("fetcher send ctx: %w", ctx.Err())
+	case f.out <- r:
+		return nil
+	}
+}
+
+func (f *FetchWorker) buildRecord(fields []string, values []any) sdk.Record {
+	payload := make(sdk.StructuredData)
+
+	for i, name := range fields {
+		switch t := values[i].(type) {
+		case time.Time: // type not supported in sdk.Record
+			payload[name] = t.UTC().String()
+		default:
+			payload[name] = t
+		}
+	}
+
+	// always coerce snapshot position to bigint, pk may be any type of integer.
+	lastRead := keyInt64(payload[f.conf.Key])
+
+	pos := position.Position{
+		Type: position.TypeSnapshot,
+		Snapshot: position.SnapshotPositions{
+			f.conf.Table: {
+				LastRead:    lastRead,
+				SnapshotEnd: f.snapshotEnd,
+				Done:        f.snapshotEnd == lastRead,
+			},
+		},
+	}.ToSDKPosition()
+
+	meta := map[string]string{
+		"table": f.conf.Table,
+	}
+
+	key := sdk.StructuredData{
+		f.conf.Key: payload[f.conf.Key],
+	}
+
+	return sdk.Util.Source.NewRecordSnapshot(pos, meta, key, payload)
+}
+
+func (f *FetchWorker) withSnapshot(ctx context.Context, tx pgx.Tx) error {
+	if f.conf.Snapshot == "" {
+		sdk.Logger(ctx).Warn().
+			Msgf("fetcher %q starting without transaction snapshot", f.cursorName)
+		return nil
+	}
+
+	if _, err := tx.Exec(
+		ctx,
+		fmt.Sprintf("SET TRANSACTION SNAPSHOT '%s'", f.conf.Snapshot),
+	); err != nil {
+		return fmt.Errorf("failed to set tx snapshot %q: %w", f.conf.Snapshot, err)
+	}
+
+	return nil
+}
+
+func (FetchWorker) validateKey(ctx context.Context, table, key string, tx pgx.Tx) error {
```

Comment:
Let's not mix pointer and value receivers please 🙏 Same for `validateTable`.

```suggestion
func (*FetchWorker) validateKey(ctx context.Context, table, key string, tx pgx.Tx) error {
```

Reply:
Oooh sure, will do.